src/proxy-animation.js

```diff
@@ -30,151 +30,470 @@ function createReadyPromise(details) {
         // Reject with a non-visible AbortError.
         reject(new DOMException("Animation aborted", "AbortError"));
       } else {
+        // Apply pending playbackRate
+        if (typeof details.pendingPlaybackRate == 'number') {
+          const previousCurrentTime = details.animation.currentTime;
+          details.animation.playbackRate = details.pendingPlaybackRate;
+          const timelineTime = details.timeline.currentTime;
+          details.startTime = details.pendingPlaybackRate ?
+              timelineTime - previousCurrentTime / details.pendingPlaybackRate :
+              previousCurrentTime;
+          details.pendingPlaybackRate = null;
+        }
+        switch(details.playState) {
+          case 'paused':
+            details.startTime = null;
+            break;
+
+          case 'running':
+          case 'finished':
+            details.holdTime = null;
+        }
         resolve();
       }
     });
   });
 }
 
+function getAnimationProperty(details, name) {
+  if (details.timeline && (name in details))
+    return details[name];
+  else
+    return details.animation[name];
+}
+
+function setAnimationProperty(details, name, value) {
+  if (details.timeline && (name in details))
+    details[name] = value;
+  else
+    details.animation[name] = value;
+}
+
+function effectivePlaybackRate(details) {
+  if (details.pendingPlaybackRate)
+    return details.pendingPlaybackRate;
+  return details.animation.playbackRate;
+}
+
+function applyPendingPlaybackRate(details) {
+  if (details.pendingPlaybackRate) {
+    details.animation.playbackRate = details.pendingPlaybackRate;
+    details.pendingPlaybackRate = null;
+  }
+}
+
+function updateFinishedState(details) {
+  if (!details.timeline)
+    return;
+
+  const isFinished = () => {
+    if (details.playState == 'paused')
+      return false;
+
+    const playbackRate = effectivePlaybackRate(details);
+    const currentTime = details.animation.currentTime;
+    if (playbackRate < 0 && currentTime <= 0)
+      return true;
+    if (playbackRate > 0 &&
+        currentTime >= details.animation.effect.getTiming().duration)
+      return true;
+
+    return false;
+  };
+
+  if (isFinished()) {
+    if (details.playState != 'finished') {
+      details.playState = 'finished';
```

Comment:
Do we need to call removeAnimation? If not, and finished animations continue ticking, we shouldn't need to call addAnimation on line 113, right?